lib/puppet/util/character_encoding.rb

```diff
@@ -13,12 +12,16 @@ class << self
     # @raise [Puppet::Error] a failure to legitimately set external encoding or
     #   transcode string
     def convert_to_utf_8(string)
-      return string if string.encoding == Encoding::UTF_8
-
-      value_to_encode = string.dup
+      return string if (string.encoding == Encoding::UTF_8 && string.valid_encoding?)
 
       begin
-        if valid_utf_8?(value_to_encode)
+        if already_invalid_utf_8?(string)
+          # If a string is currently believed to be UTF-8, but is also not
+          # valid_encoding?, we have no recourse but to fail because we have no
+          # idea what encoding this string originally came from where it *was*
+          # valid - all we know is it's not currently valid UTF-8.
+          raise EncodingError
```

Comment:
Calling out these lines specifically as a new addition for review - change to the heuristic, not necessarily related to the Etc wrapper but discovered when trying to improve it.